src/inc/til/throttled_func.h

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+namespace til
+{
+    namespace details
+    {
+        template<typename... Args>
+        class throttled_func_storage
+        {
+        public:
+            template<typename... MakeArgs>
+            bool emplace(MakeArgs&&... args)
+            {
+                std::unique_lock guard{ _lock };
+                const bool hadValue = _pendingRunArgs.has_value();
+                _pendingRunArgs.emplace(std::forward<MakeArgs>(args)...);
+                return hadValue;
+            }
+
+            template<typename F>
+            void modify_pending(F f)
+            {
+                std::unique_lock guard{ _lock };
+                if (_pendingRunArgs)
+                {
+                    std::apply(f, *_pendingRunArgs);
+                }
+            }
+
+            std::tuple<Args...> take()
+            {
+                std::unique_lock guard{ _lock };
+                auto pendingRunArgs = std::move(*_pendingRunArgs);
+                _pendingRunArgs.reset();
+                return pendingRunArgs;
+            }
+
+            explicit operator bool() const
+            {
+                std::shared_lock guard{ _lock };
+                return _pendingRunArgs.has_value();
+            }
+
+        private:
+            // std::mutex uses imperfect Critical Sections on Windows.
+            // --> std::shared_mutex uses SRW locks that are small and fast.
+            mutable std::shared_mutex _lock;
+            std::optional<std::tuple<Args...>> _pendingRunArgs;
```

Comment:
Did you consider making this use til::mutex?

Reply:
I did, but I wanted to keep both header files independent of each other and keep their template complexity minimal. Would you suggest that I use `til::shared_mutex` here?